src/naive/datetime/serde.rs

```diff
@@ -505,92 +1794,160 @@ pub mod ts_microseconds_option {
         fn visit_some<D>(self, d: D) -> Result<Self::Value, D::Error>
         where
             D: de::Deserializer<'de>,
         {
-            d.deserialize_i64(MicroSecondsTimestampVisitor).map(Some)
+            d.deserialize_str(ts_milliseconds_string::MilliSecondsTimestampVisitor).map(Some)
         }
 
-        /// Deserialize a timestamp in microseconds since the epoch
+        /// Deserialize a string timestamp in milliseconds since the epoch
         fn visit_none<E>(self) -> Result<Self::Value, E>
         where
             E: de::Error,
         {
             Ok(None)
         }
 
-        /// Deserialize a timestamp in microseconds since the epoch
+        /// Deserialize a string timestamp in milliseconds since the epoch
         fn visit_unit<E>(self) -> Result<Self::Value, E>
         where
             E: de::Error,
         {
             Ok(None)
         }
     }
+
+    #[cfg(all(test, feature = "serde"))]
+    mod tests {
+        use crate::naive::serde::ts_milliseconds_string_option;
+        use crate::{NaiveDate, NaiveDateTime};
+        use serde_derive::{Deserialize, Serialize};
+
+        #[derive(Deserialize, Serialize)]
+        struct S {
+            #[serde(with = "ts_milliseconds_string_option")]
+            time: Option<NaiveDateTime>,
+        }
+
+        #[test]
+        fn test_serialize_naive_ts_milliseconds_string_option_some() -> Result<(), serde_json::Error>
+        {
+            let my_s = S {
+                time: Some(
+                    NaiveDate::from_ymd_opt(2018, 5, 17)
+                        .unwrap()
+                        .and_hms_milli_opt(02, 04, 59, 918)
+                        .unwrap(),
+                ),
+            };
+
+            let as_string = serde_json::to_string(&my_s)?;
+
+            assert_eq!(as_string, r#"{"time":"1526522699918"}"#);
+
+            Ok(())
+        }
+
+        #[test]
+        fn test_serialize_naive_ts_milliseconds_string_option_none() -> Result<(), serde_json::Error>
+        {
+            let my_s = S { time: None };
+
+            let as_string = serde_json::to_string(&my_s)?;
+
+            assert_eq!(as_string, r#"{"time":null}"#);
+
+            Ok(())
+        }
+
+        #[test]
+        fn test_deserialize_naive_ts_milliseconds_string_option_some(
+        ) -> Result<(), serde_json::Error> {
+            const JSON: &'static str = r#"{"time":"1526522699918"}"#;
+
+            let my_s: S = serde_json::from_str(JSON)?;
+
+            assert_eq!(my_s.time.unwrap().timestamp_millis(), 1526522699918);
+
+            Ok(())
+        }
+
+        #[test]
+        fn test_deserialize_naive_ts_milliseconds_string_option_none(
+        ) -> Result<(), serde_json::Error> {
+            const JSON: &'static str = r#"{"time":null}"#;
+
+            let my_s: S = serde_json::from_str(JSON)?;
+
+            assert_eq!(my_s.time, None);
+
+            Ok(())
+        }
+    }
```

Comment:
Same question about testing maximum and minimum boundaries, and values immediately outside those boundaries.